### PR TITLE
Possible fix for issue #188

### PR DESCRIPTION
**Pull request by roveri-marco**
This is  a possible fix for issue #188 
- Now it correctly parse `(= ?r1 ?r2)`, `(= ?r rob)`, `(= (numeric_fluent ?r) (numeric_fluent2 ?r1 ?r2))` where `rob` is a constant and `?r` , `?r1`, and '?r2' are action parameters. 
-  Added supports for constants in the arguments of numeric fluents (e.g. `(numeric_fluent rob ?r)`)
-  Added some informative output for not supported features (e.g. `#t` as used in PDDL+)
-  Added one test case for the successful parsing

### CODE REVIEW

**Comment by jjzapf** [comment]
Still reviewing these changes, but over all they look good. Nice job!

The use of "==" to compare 2 doubles still seems a bit dubious to me. I'm not really sure about the best way to handle this. However, I did come across this approach, which might work.

https://en.cppreference.com/w/cpp/types/numeric_limits/epsilon

In that example, the comparison epsilon value is defined in terms of the machine epsilon value, the magnitude of the values, and the desired precision in ULPs. Perhaps we could just hard code the number of ULPs? (The example uses a value of 2.) Even with hard coding the number of ULPs, the comparison epsilon value will vary with the magnitude of the input values. This seems like it could strike the correct balance between problem generality and ease of use.

File: plansys2_pddl_parser/src/plansys2_pddl_parser/Utils.cpp [hunk 1]
```diff
@@ -14,7 +14,7 @@
 
 #include <regex>
 #include <iostream>
-#include <functional> 
+#include <functional>
```
Comment (jjzapf) [inline plansys2_pddl_parser/src/plansys2_pddl_parser/Utils.cpp:17]: Is this include needed?

File: plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp [hunk 2]
```diff
@@ -63,34 +71,90 @@ Expression * createExpression( Stringreader & f, TokenStruct< std::string > & ts
 		f.next();
 		std::string s = f.getToken();
 		if ( s == "+" || s == "-" || s == "*" || s == "/" ) {
```
Comment (jjzapf) [inline plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp:73]: Should we also check for these: "=", ">=", ">", "<=", "<" ?

File: plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp [hunk 3]
```diff
@@ -63,34 +71,90 @@ Expression * createExpression( Stringreader & f, TokenStruct< std::string > & ts
...
+   // It is a composite expression
 			CompositeExpression * ce = new CompositeExpression( s );
 			ce->parse( f, ts, d );
 			return ce;
-		}
-		else {
+		}	else {
+   // It is a function expression
```
Comment (jjzapf) [inline plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp:79]: Unfortunately, most of the plansys2_pddl_parser files seem to use tabs instead of spaces for indentation. It's a bit hard to follow all the if, else if, and else statements in these changes as I believe you are using spaces. Could you use tabs to follow the original indentation scheme?

In the future, I would support replacing all the tabs in these files with spaces.

**Comment by roveri-marco** [comment]
> 

Indeed, this could be a good approach in the internals, but at planning level while specifying the domain one is expected to use =. We may handle this case in the proper way in a second PR if ok for you, otherwise we can extend this to handle as suggested.

File: plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp (hunk 3 above)
Comment (roveri-marco) [reply plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp:79]: If ok, I can change all the tabs to spaces since easier ;-) I was not noticing since this package is not subject to lint as for instance it is the case for the terminal.

File: plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp (hunk 2 above)
Comment (roveri-marco) [reply plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp:73]: I'm not sure since here we are building expressions (e.g. `(= (* (+ a b) d)))`) and it is not possible to mix equality within expressions. The relational operators are handled above if I'm not mistaken.

File: plansys2_problem_expert/src/plansys2_problem_expert/Utils.cpp [hunk 4]
```diff
@@ -200,6 +200,14 @@ std::tuple<bool, bool, double> evaluate(
               return std::make_tuple(true, false, 0);
             }
             break;
+          case plansys2_msgs::msg::Node::COMP_EQ:
+            if (std::get<2>(left) == std::get<2>(right)) {
```
[the remark this answers is not in the document]
Comment (roveri-marco) [reply plansys2_problem_expert/src/plansys2_problem_expert/Utils.cpp:204]: Ok, we can handle it as suggested then.

File: plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp (hunk 3 above)
Comment (jjzapf) [reply plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp:79]: I'd be fine converting all tabs to spaces. Although, that might create a very large diff in this package. Perhaps it would be better to do that in a separate PR. We should see what @fmrico has to say about this.

Comment (fmrico) [reply plansys2_pddl_parser/src/plansys2_pddl_parser/Expression.cpp:79]: Yes, probably a separate PR is better. And, of, of course, spaces better tan tabs

**Comment by fmrico** [comment]
LGTM!! Merging

> We may handle this case in the proper way in a second PR if ok for you, otherwise we can extend this to handle as suggested.
:+1:
